lte/gateway/python/magma/mobilityd/ipv6_allocator_pool.py

```diff
@@ -0,0 +1,176 @@
+            removed_blocks = []
+            if self._assigned_ip_block in ipblocks:
+                # Clear allocated session prefix and IID store
+                self._allocated_iid.clear()
+                self._sid_session_prefix_allocated.clear()
+
+                removed_blocks.append(self._assigned_ip_block)
+                self._assigned_ip_block = None
+        return removed_blocks
+
+    def alloc_ip_address(self, sid: str, _) -> ip_address:
+        """
+        Calculates full IPv6 Address from configured prefix part
+        (assigned_ip_block) + unique session prefix part + interface
+        identifier
+
+        :param sid: composite SID (IMSI + apn)
+        :param _:
+        :return: Full IPv6 address
+        """
+        with self._lock:
+            # Take available ipv6 host from network
+            ipv6_addr_part = next(self._assigned_ip_block.hosts())
+
+            # Calculate session part from rest of 64 prefix bits
+            session_prefix_part = self._get_session_prefix(sid)
+
+            # Get interface identifier from 64 bits fixed length
+            iid_part = self._get_ipv6_iid_part(IID_PART_LEN)
+            if not iid_part:
+                logging.error('Could not get IPv6 IID for sid: %s', sid)
+                raise MaxRandIIDCalculationError(
+                    'Could not get IPv6 IID for sid: %s', sid)
+
+            ipv6_addr = ipv6_addr_part + (session_prefix_part * iid_part)
+            return ipv6_addr
+
+    def release_ip(self, ip_desc: IPDesc):
+        """
+        Deallocates session prefix part from identifier and allocated IID
+        :param ip_desc: IP Desc to remove
+
+        """
+        with self._lock:
+            sid = ip_desc.sid
+            ip_addr = ip_desc.ip
+            ipv6_addr_part = int(next(self._assigned_ip_block.hosts()))
+
+            session_prefix = self._sid_session_prefix_allocated.get(sid)
+            if not session_prefix:
+                raise IPNotInUseError('IP %s not allocated', ip_addr)
+
+            if ip_addr in self._assigned_ip_block and session_prefix:
+                # Extract IID part of the given IPv6 prefix and session prefix
+                iid_part = float(
+                    (int(ip_addr) - ipv6_addr_part) / int(session_prefix))
+
+                if iid_part in self._allocated_iid:
+                    del self._sid_session_prefix_allocated[sid]
+                    self._allocated_iid.remove(iid_part)
+                else:
+                    raise IPNotInUseError('IP %s not allocated', ip_addr)
+
+    def _get_ipv6_iid_part(self, length: int = 64) -> Optional[int]:
+        """
+        Calculates IPv6 Interface identifier using random calculation,
+        if it exceeds MAX_RAND calculation tries, it returns None
+        :param length: length for ipv6 IID
+        :return: Randomized IID N bits
+        """
+        for i in range(MAX_RAND_IID_CALC_TRIES):
+            rand_iid_bits = random.getrandbits(length)
```

Comment:
I assume this is seeded with system clock underneath the random class?

Reply:
that's corrrect, python random default seed will be generated using current system clock time